File: src/pipeline.js
import { loadConfig } from './utils/config.js';
import { detectProject } from './detectors/index.js';
import { buildProject } from './adapters/builders/index.js';
import { testProject } from './adapters/builders/test.js';
import { deployProject } from './adapters/deployers/index.js';

export async function runPipeline({ cwd, configPath, stage = 'all' }) {
  const cfg = await loadConfig(configPath);
  const detected = await detectProject({ cwd, cfg });

  if (stage === 'detect') {
    console.log(JSON.stringify(detected, null, 2));
    return;
  }

  const buildResult = stage === 'detect' ? null : await buildProject({ cwd, cfg, detected, skip: stage !== 'all' && stage !== 'build' ? true : false });
  
  if (stage === 'build') return;

  await testProject({ cwd, cfg, detected, skip: stage !== 'all' && stage !== 'test' ? true : false });

  if (stage === 'test') return;

  await deployProject({ cwd, cfg, detected, skip: stage !== 'all' && stage !== 'deploy' ? true : false });
}

File: automata_cli/pipeline.py
from pathlib import Path
from .utils.config import load_config
from .detectors import detect_project
from .runners.builders import build_project
from .runners.tests import test_project
from .runners.deploy import deploy_project
from .generators.config_generator import auto_generate_config


def run_pipeline(*, cwd: Path, config_path: Path, stage: str = 'all') -> None:
    # Сначала обнаруживаем языки
    detected = detect_project(cwd)
    
    # Автоматически генерируем конфиг если его нет
    if not config_path.exists():
        print("No config found, auto-generating...")
        auto_generate_config(cwd, detected, force=True)
        config_path = cwd / 'automata.yml'
    
    cfg = load_config(config_path)

    if stage == 'detect':
        import json
        print(json.dumps(detected, indent=2))
        return

    skip_build = stage not in ('all', 'build')
    skip_test = stage not in ('all', 'test')
    skip_deploy = stage not in ('all', 'deploy')

    build_project(cwd, cfg, detected, skip=skip_build)
    if stage == 'build':
        return

    test_project(cwd, cfg, detected, skip=skip_test)
    if stage == 'test':
        return

    deploy_project(cwd, cfg, detected, skip=skip_deploy)


